python/uv/private/lock.bzl

```diff
@@ -13,113 +13,424 @@
+        "_template": attr.label(
+            default = "//python/uv/private:lock_template",
+            doc = """\
+The template to be used for 'uv pip compile'. This is either .ps1 or bash
+script depending on what the target platform is executed on.
+""",
+        ),
+    },
+    executable = True,
+)
+
+def _maybe_file(path):
+    """A small function to return a list of existing outputs.
+
+    If the file referenced by the input argument exists, then it will return
+    it, otherwise it will return an empty list. This is useful to for programs
+    like pip-compile which behave differently if the output file exists and
+    update the output file in place.
+
+    The API of the function ensures that path is not a glob itself.
 
     Args:
-        name: The name of the target to run for updating the requirements.
-        srcs: The srcs to use as inputs.
-        out: The output file.
-        upgrade: Tell `uv` to always upgrade the dependencies instead of
-            keeping them as they are.
-        universal: Tell `uv` to generate a universal lock file.
-        args: Extra args to pass to the rule.
-        **kwargs: Extra kwargs passed to the binary rule.
+        path: {type}`str` the file name.
     """
-    pkg = native.package_name()
-    update_target = name + ".update"
-
-    _args = [
-        "--custom-compile-command='bazel run //{}:{}'".format(pkg, update_target),
-        "--generate-hashes",
-        "--emit-index-url",
-        "--no-strip-extras",
-        "--python=$(PYTHON3)",
-    ] + args + [
-        "$(location {})".format(src)
-        for src in srcs
-    ]
-    if upgrade:
-        _args.append("--upgrade")
-    if universal:
-        _args.append("--universal")
-    _args.append("--output-file=$@")
-    cmd = "$(UV_BIN) pip compile " + " ".join(_args)
+    for p in native.glob([path], allow_empty = True):
+        if path == p:
+            return p
+
+    return None
+
+def lock(
+        *,
+        name,
+        args = [],
+        build_constraints = [],
+        constraints = [],
+        env = None,
+        generate_hashes = True,
+        out,
+        python_version = None,
+        srcs,
+        strip_extras = False,
+        **kwargs):
+    """Pin the requirements based on the src files.
 
-    # Make a copy to ensure that we are not modifying the initial list
-    srcs = list(srcs)
+    This macro creates the following targets:
+    - `name`: the target that creates the requirements.txt file in a build
+      action.
+    - `name.run`: a runnable target that can be used to pass extra parameters
+      to the same command that would be run in the `name` action. This will
+      update the source copy of the requirements file.
+    - `name.update`: a target that can be run to update the source-tree version
+      of the requirements lock file. The output can be fed to the
+      {obj}`pip.parse` bzlmod extension tag class. Note, you can use
+      `native_test` to wrap this target to make a test.
```

Comment:
nit: the docs should better distinguish between run vs update.

Run: You can customize the args via the command line, but it requires being able to run uv (and possibly python) directly on your host.

Update: You can't customize the args via command line, but you can use RBE to generate requirements (offload execution and run for different platforms)

Reply:
Thanks for suggestions.